Plugin addition - Submit Folksonomy Tags

Viktini opened the pull request:
You may know me as Victini on MusicBrainz or Flaky on Discord.

Put briefly, this plugin lets users submit tags from their files as MusicBrainz' folksonomy tags. I've gone into more detail here: https://community.metabrainz.org/t/submit-folksonomy-tags-plugin-for-submitting-tags-to-musicbrainz-via-picard/626105

The only major bug I have with the plugin is that submitting after logging in from Picard's "to do this you must login" popup gives me an Unauthorised error (any subsequent submits are fine and of course, logging on before submitting), but Aerozol suggested on Discord that I should submit this anyhow and fix it later.

Commit:
add submit folksonomy tag plugin

Commit:
Update README.md

Commit:
removed unused variables and whitespace

Commit:
update submit_folksonomy_tags README 

based on initial checks from codacy https://github.com/metabrainz/picard-plugins/runs/12546220957

I'm new to this, so... bleh.

Commit:
another README.md fix for submit-folksonomy-tags

Codacy really doesn't like the README.md (namely the list-item indicies), fixing it again
https://github.com/metabrainz/picard-plugins/runs/12546981858

rdswift commented on plugins/submit_folksonomy_tags/__init__.py:
```diff
@@ -0,0 +1,440 @@
+    if inconsistent_detected or alert_multiple_mbids:
+        warning = QMessageBox()
+        warning.setStandardButtons(QMessageBox.Ok|QMessageBox.Cancel)
+        warning.setDefaultButton(QMessageBox.Cancel)
```
Since you use the same icon in all cases, you might consider setting it here rather than in each conditional block.

Viktini replied:
Alright, thanks!

Commit:
Remove redundant warning message lines

Viktini commented:
Only major issue is the complexity now. I should be able to work on that next week, if that's alright?

rdswift commented:
> Only major issue is the complexity now. I should be able to work on that next week, if that's alright?

That should be fine.  I don't think there's any need to rush.  The current working version is available from your repository if anybody wants to try it out, so I suggest that you take whatever time you need.  For what it's worth, I still have a tendency to include everything in the same function / module / method myself.  I think it's because I've done way too much "linear" programming in the past, and haven't totally embraced the concepts of OOP.  :grin: 

One final thing, and this is really minor...  I appreciate the credit that you've provided in the comments in your code, but I noticed that you referred to 'rswift' rather than 'rdswift'.  Like I said, really minor.  :wink:

Commit:
correct credit

Viktini commented:
Alright, thank you! :)

I've tested the plugin on Arch Linux (Python 3.10.10) / Picard 2.8.5 and all seems to be working just fine now. The complexity of the main function should still be sorted in the near future, but if there's no other problems I think it can be merged.

phw commented on plugins/submit_folksonomy_tags/__init__.py:
```diff
@@ -0,0 +1,438 @@
+        if track.files:
+            for file in track.files:
+                mbid_list = [html.escape(tag.strip().lower()) for tag
+                             in re.split(";|/|,", file.metadata[target_tag])]
```
Instead of the splitting `file.metadata.getall(target_tag)`  should be used to obtain the list of individual values.

phw commented on plugins/submit_folksonomy_tags/__init__.py:
```diff
@@ -0,0 +1,438 @@
+    # Variable to enable alert if multiple MBIDs are associated, must be toggled.
+    alert_multiple_mbids = False
+
+    # TODO when Windows Picard updates with Python 3.10, use case/switch.
```
Just a side note: It will take quite a while until we consider 3.10 the minimum supported version. Essentially this will happen when either 3.9 is no longer supported upstream or when we have some really urgent need to use something not available in earlier Python versions.

phw commented on plugins/submit_folksonomy_tags/__init__.py:
```diff
@@ -0,0 +1,438 @@
+            xml_data.extend([f'<{key} id="{mbid}">', "<user-tag-list>"])
+            # add the tags
+            for tag in data[key][mbid]:
+                xml_data.append(f'<user-tag{upvote_tag_fill}><name>{process_tag_aliases(tag.lower())}</name></user-tag>')
```
Are the tags already XML escaped at this point?

Viktini replied:
I don't think so, but I can get that sorted. Shouldn't be a problem.

phw replied:
From reading the source it is probably fine because of https://github.com/metabrainz/picard-plugins/pull/349/files#diff-9cccf3c2c55a90e5485eb300d6ad5dba86a9cb517fbd46f45f2573dce5ac62b3R200 . But please double check.

Maybe it is even better to move the escaping to here where the XML is actually build. That makes it less likely this breaks somewhere in between after some refactoring.

Viktini replied:
Before I go and change it, should I keep to the current library or move to `xml.sax.saxutils` instead?

phw replied:
General rule to follow is to stick to those modules Picard already uses, as only those are available in the binary builds for Windows and macOS.

But having said this `xml.sax.saxutils` should be fine and is probably the best choice here. Picard already uses this internally at https://github.com/metabrainz/picard/blob/master/picard/webservice/api_helpers.py

Just as a side note: There is not much XML functionality left in Picard as all the core services moved to JSON. The ratings submission is the only built-in place left where XML is actively used. Apart from that only https://github.com/metabrainz/picard/blob/master/picard/util/xml.py exists for parsing XML (based on Qt functionality), but it's only there for use by plugins that access XML web services.

phw reviewed (changes requested):
Thanks a lot for submitting this. Overall looks good to me and I had already tested your plugin a while back. Only one suggestion to use `metadata.getall`  instead of splitting tag values.

> The complexity of the main function should still be sorted in the near future, but if there's no other problems I think it can be merged.

Agreed. Factoring out some code into separate functions should do the trick, but I'm fine with merging this as is for now.

Commit:
Change + move escape, use getall to get MBIDs 

uses `xml.sax.saxutils` instead of `html` to provide XML escaping, and move said escaping into the function for generating XMLs instead of handling tags. I've also used `.getall()` for grabbing MBIDs as phw suggested.

Viktini commented:
I've started work on refactoring, and thus far I've managed to split the submission handler into different functions in a class. I'm using Radon as a way to check if I'm on the right track, will that pose an issue with Codacy?
